aten/src/ATen/native/AutogradComposite.cpp

```diff
@@ -22,6 +22,54 @@ std::tuple<at::Tensor, at::Tensor> _unpack_dual(const at::Tensor& tensor, int64_
   return std::tuple<at::Tensor, at::Tensor>(tensor._fw_primal(level), tensor._fw_grad(level));
 }
 
+Tensor _new_zeros_with_same_feature_meta(
+    const at::Tensor& self,
+    const at::Tensor& other,
+    c10::optional<int64_t> self_num_batch_dims) {
+  // NB: This function can be called directly from _set_fw_grad or
+  //     if self is batched, from this function's batching rule.
+  auto other_sizes = other.sizes();
+  auto other_strides = other.strides();
+  auto other_storage_offset = other.storage_offset();
+  int64_t other_storage_numel = other.storage().nbytes() / other.itemsize();
+
+  if (!self_num_batch_dims.has_value()) {
+    auto new_tensor = at::zeros({other_storage_numel}, other.options());
+    return new_tensor.as_strided(other_sizes, other_strides, other_storage_offset);
+  }
+
+  int64_t _self_num_batch_dims = self_num_batch_dims.value();
+
+  auto self_sizes = self.sizes();
+  auto self_strides = self.strides();
+
+  // NB: We don't check that the sizes of self is the same as that of other
+  //     because this function is also used in the inplace over view case
+
+  std::vector<int64_t> out_sizes;
+  out_sizes.reserve(other.dim() + _self_num_batch_dims);
+  out_sizes.insert(out_sizes.end(), self_sizes.begin(), self_sizes.begin() + _self_num_batch_dims);
+  out_sizes.insert(out_sizes.end(), other_sizes.begin(), other_sizes.end());
```

Comment:
nit: use SmallBuffer for better performance and also because we don't need to re-size the allocation. Ditto for out_strides.